### ISSUE TITLE
Project build order

### ISSUE THREAD

**Issue by freddydk** [microsoft/AL-Go#264]
### Summary

Currently, if you have a multi-project repository with projects, where you have dependencies between the projects, then you cannot create a PR, which covers multiple projects and have them built in the right order.

Example:
![Image](https://user-images.githubusercontent.com/10775043/197638160-05e68f41-9e2e-4c13-a27e-2f5c4952d888.png)

If you checkin some code, which introduces an event in the Common app AND uses it in App.DK - then the projects will be built simultaneously and App.DK will likely fail, because the version of its dependency App.Common, that job will download from the latest build, doesn't contain the new event yet.

This feature will fix this behavior.

### Intended Outcome

Running Update AL-Go System files will detect dependencies between projects and automatically determine the best project build order.
In the above picture, the order would be:
1. Project Common
2. Project App.WorldWide
3. Project App.Denmark and App.Italy simultanously

### How it works

A new setting UseProjectDependencies will activate this new behavior if set to true.
New projects will automatically get this property set to true.

The Update AL-Go System files action will build a dependency graph between the projects and setup multiple jobs in the build workflows to build things in the right order.
A new project added to the repo will be built together with the last group.
A new project added to the repo, will cause you to get warnings that there are updates to AL-Go System files

**Comment by gntpet** [comment]
hey @freddydk, 

I went through this issue/documentation regarding 'useProjectDependencies'. Maybe I missuderstand it, or i don't see full picture, but I think it's not what it's needed for mult-root projects. With multi-root projects, dependencies are discovered and built for each project individually. 

Our scenario is: 

- have multi-root project in single repo  
- appDependencyProbingPaths points to the same repo 
- release_status = include
- we do have core app SRS.core, and two country apps SRS.it, and SRS.no, which depdends on SRS.core

This setup covers the scenario you describe: 

> "If you checkin some code, which introduces an event in the Common app AND uses it in App.DK - then the projects will be built simultaneously and App.DK will likely fail, because the version of its dependency App.Common, that job will download from the latest build, doesn't contain the new event yet.

Compilation works perfectly, since every project builds it's needed dependencies. And since it's using matrix build, all projects are built in parallel, which is super fast compared to sequential builds.

But this approach is not without a problem. You do have a code which checks, which files got changed, and just build only those projects without taking dependencies into account.

For e.g. if you change files only SRS.core project, it will build SRS.core only. And projects which depends on SRS.core (for e.g SRS.it), will not be rebuilt. So, if I do a breaking change in SRS.core (changing procedure from public to internal), my SRS.Core will be built successfully, will be deployed to QA environment, and user will get runtime error, while running my private function called from SRS.IT.

There two solutions:

1. always build all projects
2. use dependency graph to determine if other projects needs to be rebuild as well

Here's an deployment example, where u see that only SRS.core got compiled/deployed.
![image](https://user-images.githubusercontent.com/16342128/199502269-f9428474-684b-4352-9487-c8b7439ec137.png)

**Comment by freddydk** [comment]
As I read your scenario, that is exactly what UseProjectDependencies will solve.
Just set the setting in .github/AL-Go-Settings.json and remove all your appDependencyProbingPaths - then run Update AL-Go System files - and run CI/CD.
Now checkin a fix in SRS.IT - then that is the only project built.
Now checkin a fix in SRS.core - then that is built - and after that IT and NO - using the output from the core build.

**Comment by freddydk** [comment]
Said in another way - it does exactly build a project dependency graph and use that.
This is part of initialize in my buildorder project:
![image](https://user-images.githubusercontent.com/10775043/199512038-2edbe200-b996-443b-aab5-d13175e29b5e.png)

**Comment by freddydk** [comment]
and... the text you refer to:

"If you checkin some code, which introduces an event in the Common app AND uses it in App.DK - then the projects will be built simultaneously and App.DK will likely fail, because the version of its dependency App.Common, that job will download from the latest build, doesn't contain the new event yet.

Is how it worked before - this feature fixes this issue.

**Comment by gntpet** [comment]
> The problem with using include is, that you spend time building the core app in both projects - and you have a different version of the core app in each project.

- This is still way faster (3x) than running 3 build in sequence. 
- Core app is the same across all projects since it has the same code. 
- Core is compiled against different versions (base, Italy, ...). In your case, you are not handling this scenario. We experienced quite some troubles with our Core solution, especially with pageextensions and missing/different anchors. Worked perfectly fine with base, but failed with italian version.
- Yes, you generate 3 app files for Core. Their names matches, but content differs a bit. Did comparison & saw difference in one file only - **navigation.xml** (not sure, what's the purpose of it)

For us, it would be perfect a bit from both solutions. Do build as it is done now using matrix model, but check ProjectDependencies before the build.

I hope it's valuable input for your development, @freddydk

**Comment by freddydk** [comment]
The problem is that the package id in the core app is different.
That is not a problem for you, since the IT and NO version are never installed on the same customer - but for other apps - that is a problem.
I could probably use the dependency graph for the include scenario as well - then you have best of both worlds - will check that.

**Comment by freddydk** [comment]
Could you try the following...
Running on preview - try to set the UseProjectDependencies to true - and do NOT run the update al-go-system files.
Now try to modify your srs.core - and see if it builds both the dependent projects?
It should.

If you update the AL-Go System files now - then you get the new behavior, which you don't want - I will create a different setting for this "include"-behavior

**Comment by freddydk** [comment]
The above won't do what you are looking for - but it should detect projects.
They should however be combined in a different way.
In the above scenario - you would have two jobs - IT and DK - and all their dependencies would be included.
Will add another slice for that.

**Comment by gntpet** [comment]
hey @freddydk , 
I just tested it, and it worked. I made single change with SRS.core and it triggered jobs for other countries as well

**Comment by freddydk** [comment]
OK - so, I think I know what to do - and it makes a ton of sense.
I added a new slice to the backlog - and closed this one.
